google/cloud/storage/internal/grpc/metrics_meter_provider_test.cc

```diff
@@ -0,0 +1,252 @@
+  EXPECT_CALL(*mock, Export(MetricDataNotEmpty()))
+      .Times(::testing::AtLeast(1))
+      .WillRepeatedly(
+          [](opentelemetry::sdk::metrics::ResourceMetrics const& data) {
+            EXPECT_THAT(
+                data.scope_metric_data_,
+                Contains(WithMetricsData(Contains(AllOf(
+                    MatchesInstrumentName(kDurationMetric),
+                    WithPointData(ElementsAre(ExpectedLatencyHistogram())))))));
+            return opentelemetry::sdk::common::ExportResult::kSuccess;
+          });
+
+  // use a new scope to force a flush from the meter and provider before the
+  // function exit.  Otherwise the mocks may not be called in time.
+  {
+    auto provider = MakeGrpcMeterProvider(std::move(mock), TestReaderOptions());
+    auto meter = provider->GetMeter("grpc-c++", grpc::Version());
+    auto histogram = meter->CreateDoubleHistogram(kDurationMetric,
+                                                  "test-only-description", "s");
+    // It may take several attempts before the periodic reader exports any data.
+    // A short loop does the trick.
+    for (int i = 0; i != 50; ++i) {
+      histogram->Record(1.0, opentelemetry::context::Context{});
+      std::this_thread::sleep_for(kExportInterval);
+    }
+  }
+}
+
+TEST(GrpcMetricsExporter, ValidateGrpcClientAttemptSize) {
+  for (std::string const size_metric : {kRcvdSizeMetric, kSentSizeMetric}) {
+    SCOPED_TRACE("Testing with " + size_metric);
+    auto mock = std::make_unique<MockPushMetricExporter>();
+    EXPECT_CALL(*mock, Shutdown).WillOnce(Return(true));
+    EXPECT_CALL(*mock, GetAggregationTemporality)
+        .WillRepeatedly(Return(
+            opentelemetry::sdk::metrics::AggregationTemporality::kCumulative));
+    // The exporter may export some empty data, we just ignore those.
+    EXPECT_CALL(*mock, Export(MetricDataEmpty()))
+        .WillRepeatedly(
+            Return(opentelemetry::sdk::common::ExportResult::kSuccess));
+    // We use a different condition on the call to validate the real data.
+    EXPECT_CALL(*mock, Export(MetricDataNotEmpty()))
+        .Times(::testing::AtLeast(1))
+        .WillRepeatedly(
+            [size_metric](
+                opentelemetry::sdk::metrics::ResourceMetrics const& data) {
+              EXPECT_THAT(
+                  data.scope_metric_data_,
+                  Contains(WithMetricsData(Contains(AllOf(
+                      MatchesInstrumentName(size_metric),
+                      WithPointData(ElementsAre(ExpectedSizeHistogram())))))));
+              return opentelemetry::sdk::common::ExportResult::kSuccess;
+            });
+
+    auto provider = MakeGrpcMeterProvider(std::move(mock), TestReaderOptions());
+    auto meter = provider->GetMeter("grpc-c++", grpc::Version());
+    auto histogram = meter->CreateDoubleHistogram(
+        size_metric, "test-only-description", "By");
+    // It may take several attempts before the periodic reader exports any data.
```

Comment:
This seems like it could be flakey. Why did you pick 50? Is it guarenteed to finish?

Reply:
If there's no other way to mock the passage of time with a clock, then I guess this should be fine. Thanks for adding the comment

Reply:
`opentelemetry-cpp` doesn't expose a clock interface for this thing. They just sleep in their tests.

https://github.com/open-telemetry/opentelemetry-cpp/blob/ce14bf6520a059ff0bca81df45603722d965ad7c/sdk/test/metrics/periodic_exporting_metric_reader_test.cc#L14

We can't do any better. :shrug: